InvenTree/machine/registry.py

```diff
@@ -0,0 +1,142 @@
+        self.discover_drivers()
+        self.load_machines()
+
+    def discover_machine_types(self):
+        import InvenTree.helpers
+
+        logger.debug("Collecting machine types")
+
+        machine_types: Dict[str, BaseMachineType] = {}
+        base_drivers: List[BaseDriver] = []
+
+        discovered_machine_types: List[BaseMachineType] = InvenTree.helpers.inheritors(BaseMachineType)
+        for machine_type in discovered_machine_types:
+            try:
+                machine_type.validate()
+            except NotImplementedError as error:
+                self.errors.append(error)
+                continue
+
+            machine_types[machine_type.SLUG] = machine_type
+            base_drivers.append(machine_type.base_driver)
+
+        self.machine_types = machine_types
+        self.base_drivers = base_drivers
+
+        logger.debug(f"Found {len(self.machine_types.keys())} machine types")
+
+    def discover_drivers(self):
+        import InvenTree.helpers
+
+        logger.debug("Collecting machine drivers")
+
+        drivers: Dict[str, BaseDriver] = {}
+
+        discovered_drivers: List[BaseDriver] = InvenTree.helpers.inheritors(BaseDriver)
+        for driver in discovered_drivers:
+            # skip discovered drivers that define a base driver for a machine type
+            if driver in self.base_drivers:
+                continue
+
+            try:
+                driver.validate()
+            except NotImplementedError as error:
+                self.errors.append(error)
+                continue
+
+            drivers[driver.SLUG] = driver
+
+        self.drivers = drivers
+
+        logger.debug(f"Found {len(self.drivers.keys())} machine drivers")
+
+    def get_driver_instance(self, slug: str):
+        if slug not in self.driver_instances:
+            driver = self.drivers.get(slug, None)
+            if driver is None:
+                return None
+
+            self.driver_instances[slug] = driver()
+
+        return self.driver_instances.get(slug, None)
+
+    def load_machines(self):
+        # Imports need to be in this level to prevent early db model imports
+        from machine.models import Machine
+
+        for machine in Machine.objects.all():
+            self.machines[machine.pk] = machine
+
+        # initialize machines after all machine instances were created
+        for machine in self.machines.values():
+            machine.initialize()
+
+    def get_machines(self, **kwargs):
+        """Get loaded machines from registry. (By default only active machines)
+
+        Kwargs:
+            name: Machine name
+            machine_type: Machine type definition (class)
+            driver: Machine driver (class)
+            active: (bool, default: True)
+            base_driver: base driver (class | List[class])
+        """
+        allowed_fields = ["name", "machine_type", "driver", "active", "base_driver"]
+
+        kwargs = {**{'active': True}, **kwargs}
```

Comment:
Question: is there a reason to not put this into the function definition with a default?

Reply:
Because then the generic loop does not work that way?